tools/ktx/command_compare.cpp

```diff
@@ -0,0 +1,2465 @@
+
+                printIndent(0, "{}", nl);
+                printIndent(3, "}}");
+            }
+            if (diff.texelBlockPairList.empty()) {
+                printIndent(0, "]");
+            } else {
+                printIndent(0, "{}", nl);
+                printIndent(2, "]");
+            }
+        }
+    }
+};
+
+// -------------------------------------------------------------------------------------------------
+
+/** @page ktx_compare ktx compare
+@~English
+
+Compare two KTX2 files.
+
+@section ktx_compare_synopsis SYNOPSIS
+    ktx compare [option...] @e input-file1 @e input-file2
+
+@section ktx_compare_description DESCRIPTION
+    @b ktx @b compare compares the two KTX2 files specified as the @e input-file1 and @e input-file2
+    arguments and outputs any mismatch in texture information and/or image data.
+    The command implicitly calls @ref ktx_validate "validate" and prints any found errors
+    and warnings to stdout.
+    If any of the specified input files are invalid then comparison is done based on best effort
+    and may be incomplete.
+
+    The JSON output formats conform to the https://schema.khronos.org/ktx/compare_v0.json
+    schema even if the input file is invalid and certain information cannot be parsed or
+    displayed.
+    Additionally, for JSON outputs the KTX file identifier is printed using "\u001A" instead of
+    "\x1A" as an unescaped "\x1A" sequence inside a JSON string breaks nearly every JSON tool.
+    Note that this does not change the value of the string only its representation.
+
+    @note @b ktx @b compare prints using UTF-8 encoding. If your console is not
+    set for UTF-8 you will see incorrect characters in output of the file
+    identifier on each side of the "KTX nn".
+
+    The following options are available:
+    @snippet{doc} ktx/command.h command options_format
+    <dl>
+        <dt>\--content raw | image | ignore</dt>
+        <dd>Controls how image content is compared. Possible options are: <br />
+            @b raw - Encoded image data is compared verbatim, as it appears in the file. <br />
+            @b image - Effective image data is compared per texel block. <br />
+            @b ignore - Ignore image contents. <br />
+            The default mode is @b raw, meaning that the encoded image data must match exactly.
+            Note: When the mode is set to @b image, the effective image data of individual mip
+            levels, layers, and faces is compared texel block by texel block after applying
+            one or more of the following transformations, if needed: <br />
+            \- If the texture is supercompressed (e.g. using Zstandard or ZLIB), the images
+               are deflated before comparison. <br />
+            \- If the texture is encoded using BasisLZ, the images are transcoded to
+               R8G8B8A_UNORM before comparison. <br />
+            For block compressed textures and textures encoded using UASTC the texel blocks are
```

Comment:
I think UASTC images should be transcoded too. Then people can compare UASTC and BasisLZ compression after the fact.

Reply:
It was an explicit request from @lexaknyazev to treat them as block-compressed formats.

BasisLZ needs transcoding (something that we suggested adding support for), because it does not have a block layout, as the supercompression is part of the universal data format itself, hence it is an exception.

Reply:
Then I am misunderstanding "effective image data". What is the difference between `raw` and `image` then for block-compressed textures or UASTC?

Reply:
Raw is not doing per texel block comparison, but a simple memcmp of the entire image payload. Image comparison does per texel block comparison and for UASTC the texel blocks are individual UASTC compressed blocks.

Reply:
I thought ""effective image data" was the decoded data where possible so images of different but compatible formats could be compared. That is something we can possibly add later.

So the user visible difference is that you'll get told which layer, levels and faces differ instead of just that he images differ. Correct?

Reply:
> I thought ""effective image data" was the decoded data where possible so images of different but compatible formats could be compared. That is something we can possibly add later.

Not in the case of the current image comparison mode. Besides this UASTC specific requirement, a general requirement for this comparison mode was to compare the packed data values, because we wanted to see differences, even if the decoded channel values otherwise match (e.g. two different representations of NaNs for floats, or two different representations of UASTC blocks that otherwise lead to the same transcoded/decoded values).

For a comparison mode that only care about the "color" values (potentially with tolerance) that may also allow comparing different formats (which is not currently supported), it is expected that we'd need to transcode (and decode GPU BC formats, for example), and do the comparison of the decoded channel values instead (potentially even with swizzling applied, if needed). But that's something to decide in the future.

Currently, the two modes are: raw, to compare image data as encoded in the file, and image, to compare packed data elements on a per texel block basis.

> So the user visible difference is that you'll get told which layer, levels and faces differ instead of just that he images differ. Correct?

Yes, for raw you'll only get image-level difference reports on a per layer/level/face basis.